File: Practica 5/MF.py
import random
import time

class Articulo:
    def __init__(self, peso, valor):
        self.peso = peso
        self.valor = valor
        self.ratio = valor / peso if peso > 0 else 0

def mochila_fraccionaria(capacidad, articulos):

    # Ordenar artículos por ratio valor/peso en orden descendente
    articulos_ordenados = sorted(articulos, key=lambda x: x.ratio, reverse=True)
    
    valor_total = 0.0
    peso_actual = 0.0
    
    for articulo in articulos_ordenados:
        if peso_actual + articulo.peso <= capacidad:
            # Tomar el artículo completo
            peso_actual += articulo.peso
            valor_total += articulo.valor
        else:
            # Tomar fracción del artículo
            peso_restante = capacidad - peso_actual
            fraccion = peso_restante / articulo.peso
            valor_total += articulo.valor * fraccion
            peso_actual = capacidad
            break
    
    return valor_total, peso_actual

def generar_articulos(n, seed=42):
    """Genera n artículos con pesos y valores aleatorios"""
    random.seed(seed)
    articulos = []
    for i in range(n):
        peso = random.uniform(1, 50)
        valor = random.uniform(10, 100)
        articulos.append(Articulo(peso, valor))
    return articulos

def ejecutar_pruebas():
    """Ejecuta pruebas con 10, 100 y 1000 artículos"""
    capacidades = [250, 2500, 25000]  # Capacidades proporcionales
    cantidades = [10, 100, 1000]
    
    for i, n in enumerate(cantidades):
        print(f"\nPRUEBA CON {n} ARTÍCULOS")
        
        # Generar artículos
        articulos = generar_articulos(n, seed=42+i)
        capacidad = capacidades[i]
        
        print(f"Capacidad de la mochila: {capacidad:.2f}")
        
        # Resolver el problema
        inicio = time.time()
        valor_optimo, peso_usado = mochila_fraccionaria(capacidad, articulos)
        tiempo = time.time() - inicio
        
        print(f"Valor máximo obtenido: {valor_optimo:.2f}")
        print(f"Peso utilizado: {peso_usado:.2f} / {capacidad:.2f}")
        print(f"Tiempo de ejecución: {tiempo*1000:.4f} ms")

if __name__ == "__main__":
    ejecutar_pruebas()
    

File: MF.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

typedef struct {
    double peso;
    double valor;
    double ratio;
} Articulo;

// Función de comparación para qsort (orden descendente por ratio)
int comparar_ratio(const void *a, const void *b) {
    Articulo *art1 = (Articulo *)a;
    Articulo *art2 = (Articulo *)b;
    
    if (art2->ratio > art1->ratio) return 1;
    if (art2->ratio < art1->ratio) return -1;
    return 0;
}

// Genera número aleatorio entre min y max
double random_double(double min, double max) {
    return min + (max - min) * ((double)rand() / RAND_MAX);
}

// Genera n artículos con valores aleatorios
void generar_articulos(Articulo *articulos, int n) {
    for (int i = 0; i < n; i++) {
        articulos[i].peso = random_double(1.0, 50.0);
        articulos[i].valor = random_double(10.0, 100.0);
        articulos[i].ratio = articulos[i].valor / articulos[i].peso;
    }
}

// Resuelve el problema de la mochila fraccionaria
double mochila_fraccionaria(double capacidad, Articulo *articulos, int n, double *peso_usado) {
    // Ordenar artículos por ratio valor/peso (descendente)
    qsort(articulos, n, sizeof(Articulo), comparar_ratio);
    
    double valor_total = 0.0;
    *peso_usado = 0.0;
    
    for (int i = 0; i < n; i++) {
        if (*peso_usado + articulos[i].peso <= capacidad) {
            // Tomar el artículo completo
            *peso_usado += articulos[i].peso;
            valor_total += articulos[i].valor;
        } else {
            // Tomar fracción del artículo
            double peso_restante = capacidad - *peso_usado;
            double fraccion = peso_restante / articulos[i].peso;
            valor_total += articulos[i].valor * fraccion;
            *peso_usado = capacidad;
            break;
        }
    }
    
    return valor_total;
}

void ejecutar_prueba(int n, double capacidad, int seed) {
    printf("\nPRUEBA CON %d ARTICULOS\n", n);
    
    // Asignar memoria para artículos
    Articulo *articulos = (Articulo *)malloc(n * sizeof(Articulo));
    if (articulos == NULL) {
        printf("Error: No se pudo asignar memoria\n");
        return;
    }
    
    // Generar artículos
    srand(seed);
    generar_articulos(articulos, n);
    
    printf("Capacidad de la mochila: %.2f\n", capacidad);
    
    
    // Resolver el problema y medir tiempo
    clock_t inicio = clock();
    double peso_usado;
    double valor_optimo = mochila_fraccionaria(capacidad, articulos, n, &peso_usado);
    clock_t fin = clock();
    
    double tiempo_ms = ((double)(fin - inicio) / CLOCKS_PER_SEC) * 1000.0;
    
    
    printf("Valor máximo obtenido: %.2f\n", valor_optimo);
    printf("Peso utilizado: %.2f / %.2f\n", peso_usado, capacidad);
    printf("Tiempo de ejecución: %.4f ms\n", tiempo_ms);
    
    free(articulos);
}

int main() {

    // Pruebas con diferentes cantidades de artículos
    ejecutar_prueba(10, 250.0, 42);
    ejecutar_prueba(100, 2500.0, 43);
    ejecutar_prueba(1000, 25000.0, 44);
    
    return 0;
}
